src/box/box.cc

```diff
@@ -2958,13 +2958,24 @@ box_demote(void)
 	if (!is_box_configured)
 		return 0;
 
-	/* Currently active leader is the only one who can issue a DEMOTE. */
-	bool is_leader = txn_limbo_replica_term(&txn_limbo, instance_id) ==
-			 box_raft()->term && txn_limbo.owner_id == instance_id;
-	if (box_election_mode != ELECTION_MODE_OFF)
-		is_leader = is_leader && box_raft()->state == RAFT_STATE_LEADER;
-	if (!is_leader)
-		return 0;
+	const struct raft *raft = box_raft();
+	if (box_election_mode == ELECTION_MODE_OFF) {
+		assert(raft->state == RAFT_STATE_FOLLOWER);
+		if (raft->leader != REPLICA_ID_NIL) {
+			diag_set(ClientError, ER_NOT_LEADER, raft->leader);
+			return -1;
+		}
+		if (txn_limbo.owner_id == REPLICA_ID_NIL)
```

Comment:
In the discussion [here](https://github.com/tarantool/tarantool/issues/6860#issuecomment-1663432137) I meant to skip the term check but not the owner check.

I suppose you removed the owner check because the actual limbo owner might be dead. But do we really want to allow any node to demote the possibly running queue owner?

Maybe leave the crutch with 'promote + demote' for the case when the queue owner is dead?

Reply:
The problem is that if you were an owner in an old term, you have no guarantees in the new term. And we don't even track the limbo owner. We only track the election leader. So we can't tell if the queue is owned by anybody else in the newest term.

You said *"Not checking the owner means you may "steal" the synchro queue from the current owner."*, but it still can happen with checking the owner. Because the fact that you were an owner in term N doesn't tell anything about the owner in the term N+1.

Do I miss anything?

Reply:
Oh, I think I understood. Yes, indeed, there is no point in checking limbo owner for an old term (when synchro term is N, and election term is N + 1).

My point was that when the limbo term is up to date (equal to election term), it's better to check the limbo owner. Otherwise the running owner may be disrupted by anyone who issues a DEMOTE. And issuing demote on all replicaset nodes simultaneously will make all of them write a DEMOTE to WAL and thus lead to split-brain.

When the limbo term is outdated, it's fine to not check the owner. Even if there is another owner in term N + 1, our demote from term N will lead to split-brain and won't disrupt the owner in term N + 1.

Reply:
> What do you think?

Agreed, done.